File: common-lib/src/com/bigcorp/common/rules/impl/SpecialClientsRule.java
package com.bigcorp.common.rules.impl;

import com.bigcorp.common.rules.Rule;
import com.bigcorp.common.rules.RuleContext;
import com.bigcorp.common.model.Client;

/**
 * Special handling for specific client accounts.
 * 
 * This rule exists because certain clients have "special arrangements" 
 * that were negotiated by sales and communicated via email, not through
 * any formal process. The rules are hardcoded here because "it's just
 * a few clients" and "we'll put it in the database later."
 * 
 * That was in 1999. It's now 2002 and we have 12 special cases.
 * 
 * @author multiple
 * @since 1.0
 */
public class SpecialClientsRule implements Rule {

    public String getName() {
        return "SpecialClients";
    }

    public int getPriority() {
        return 50; // runs after tier and max value checks
    }

    public boolean isActive() {
        return true;
    }

    public boolean evaluate(RuleContext context) {
        // this rule always passes - it just adjusts things
        return true;
    }

    public void execute(RuleContext context) {
        Client client = context.getClient();
        if (client == null) return;

        String clientId = client.getClientId();

        // Henderson Capital - no commission on first 1000 shares per day
        // (negotiated by Jim in sales, 1999)
        if ("C002".equals(clientId)) {
            context.setAttribute("commission_override", Double.valueOf(0.0));
            context.addMessage("Henderson Capital: zero commission applied");
        }

        // Acme Trading - can trade 10 min before market open
        // (added per email from VP of Sales, 2000-03-15)
        if ("C001".equals(clientId)) {
            context.setAttribute("early_access", Boolean.TRUE);
            context.addMessage("Acme Trading: early market access granted");
        }

        // MegaFund - always gets PLATINUM pricing even though they're GOLD tier
        // (per verbal agreement with CEO, do not change without checking with Larry)
        if ("C004".equals(clientId)) {
            context.setAttribute("pricing_tier_override", Client.TIER_PLATINUM);
            context.addMessage("MegaFund: PLATINUM pricing override applied");
        }

        // NOTE: Pinnacle Investments (C005) was supposed to be added here
        // for reduced fees but the sales person left before documenting
        // what "reduced fees" meant. Ticket JIRA-3401 is open for this.
    }
}

File: rewrite-experiment/hypothesis-b/src/rules/impl/SpecialClientsRule.ts
import { BaseRule } from '../Rule';
import { RuleContext, setAttribute } from '../../domain/RuleContext';
import type { RuleResult } from '../Rule';

// RUL-016: Business — per-client overrides
// Priority 50, Category: Business, Behavior: PASS
// Always passes. Applies per-client overrides on execute().
// TODO: move to database (JIRA-7200)

interface SpecialClientConfig {
  commissionOverride?: number;
  pricingTierOverride?: string;
  earlyAccess?: boolean;
  multiCurrencyPriority?: boolean;
}

const SPECIAL_CLIENTS: Record<string, SpecialClientConfig> = {
  C001: { earlyAccess: true },
  C002: { commissionOverride: 0.0 },
  C003: { commissionOverride: 0.01 },
  C004: { pricingTierOverride: 'PLATINUM' },
  C005: { commissionOverride: 0.01 },
  C006: { commissionOverride: 0.0, earlyAccess: true },
  C007: { pricingTierOverride: 'PLATINUM' },
  C008: { commissionOverride: 0.005 },
  C009: { pricingTierOverride: 'GOLD' },
  C010: { commissionOverride: 0.0, multiCurrencyPriority: true },
};

export class SpecialClientsRule extends BaseRule {
  name = 'SpecialClientsRule';
  priority = 50;
  category = 'Business';
  failOpen = true;

  evaluate(_ctx: RuleContext): RuleResult {
    return 'PASS'; // Always passes
  }

  execute(ctx: RuleContext): void {
    const clientId = ctx.order.clientId;
    const config = SPECIAL_CLIENTS[clientId];

    if (!config) return;

    if (config.commissionOverride !== undefined) {
      setAttribute(ctx, 'commission_override', config.commissionOverride);
    }
    if (config.pricingTierOverride) {
      setAttribute(ctx, 'pricing_tier_override', config.pricingTierOverride);
    }
    if (config.earlyAccess) {
      setAttribute(ctx, 'early_access', true);
    }
    if (config.multiCurrencyPriority) {
      setAttribute(ctx, 'multi_currency_priority', true);
    }
  }
}
